docs/_includes/Statements.md

```diff
@@ -882,11 +898,15 @@ conditions that Dafny needs to verify when using a `decreases` expression:
 * that it is bounded.
 
 That is, the expression must strictly decrease in a well-founded ordering
-(cf. Section [Well-Founded Orders](#sec-well-founded-orders).
+(cf. Section [Well-Founded Orders](#sec-well-founded-orders)).
 
 Many times, an integral value (natural or plain integer) is the quantity
-that decreases, but other things that can be used as well. In the case of
-integers, the bound is assumed to be zero. For example, the following is
+that decreases, but other values can be used as well. In the case of
+integers, the bound is assumed to be zero.  The lower bound requires that
+when the `decreases` expression is evaluated just prior to the loop test,
+the value must be at least as large as the lower bound.
```

Comment:
This isn't exactly right. For example,

``` dafny
method Loop(from: int, to: int) {
  var i := from;
  while i <= to
    decreases to - i
  {
    i := i + 1;
  }
}
```

is correct and does not produce any errors. Yet, "when the `decreases` expression is evaluated just prior to the loop test", the value is `-1`.

The rule is that when the "back edge" is taken, the current `decreases` expression must be strictly below (in a well-founded order that Dafny predefines) the `decreases` expression at the beginning of that same iteration. The part you mention about the lower bound can then be explained by describing the well-founded order itself. (The error messages pertaining to integer/real orders are specialized so that they sometimes talk about "below 0".)

Here is the order for common types:

| type of `X` and `x`    | `x` strictly below `X`   |
|:-----------------------|:---------------------------------------------------------------|
| `bool`                 | `X && !x`                                                      |
| `int`                  | `x < X && 0 <= X`                                              |
| `real`                 | `x <= X - 1.0 && 0.0 <= X`                                     |
| `set<T>`               | `x` is a proper subset of `X`                                  |
| `multiset<T>`               | `x` is a proper multiset-subset of `X`                                  |
| `seq<T>`               | `x` is a consecutive proper sub-sequence of `X`                |
| `map<K, V>`               | `x.Keys` is a proper subset of `X.Keys`                |
| inductive datatypes    | `x` is structurally included in `X`                            |
| reference types | `x == null && X != null` |
| co-inductive datatypes | `false` |
| type parameter | `false` |

For arrow types, I _think_ the current order is essentially `false`, which should be improved. (It's tricky.)

Also, there are a few relations between the rows in the table above. For example, a datatype value `x` sitting inside a set that sits inside another datatype value `X` is considered to be strictly below `x`. Here's an illustration of that order, in a program that verifies:

``` dafny
datatype D = D(s: set<D>)

method TestD(dd: D) {
  var d := dd;
  while d != D({})
    decreases d
  {
    var x :| x in d.s;
    d := x;
  }
}
```

Reply:
Fixed the text. Added this table and discussion to the section on well-formed orders